source/clear-linux/tutorials/autofdo.rst

```diff
@@ -0,0 +1,320 @@
+
+   }
+
+   int main(){
+
+       start();
+
+       sort_array();
+
+       stop();
+
+       return 0;
+
+   }
+
+After compiling and executing this simple code, we have a baseline for
+incoming improvements to control optimizations. For example:
+
+.. code-block:: console
+
+   # gcc sort.c -o sort
+
+   # ./sort
+
+   Bubble sorting array of 30000 elements
+
+   3720 ms
+
+Basic Optimization Options
+==========================
+
+These options control various sorts of optimizations: `-O1`, `-O2`, and
+`-O3`. The GCC Optimize-Options section provides an excellent explanation.
+For our example, we will use the `-O3` option. `-O3` turns on all
+optimizations specified by -O2 plus the following options:
+
+* finline-functions
+* funswitch-loops
+* fpredictive-commoning
+* fgcse-after-reload
+* ftree-loop-vectorize
+* ftree-loop-distribute-patterns
+* ftree-slp-vectorize
+* fvect-cost-model
+* ftree-partial-pre
+* fipa-cp-clone
+
+We applying this flag to our example code and run it:
+
+.. code-block:: console
+
+   # gcc -O3 sort.c -o sort_optimized
+   #./sort_optimized
+
+   Bubble sorting array of 30000 elements
+
+   1500 ms
+
+We see the execution reduced by 59.6 percent. An impressive performance
+increase for a single optimization flag. Now, let us take into consideration
+that this optimization is based on only the static analysis of the code. The
+execution time provides no input that can tell us how the code is behaving
+for the user: which parts are never executed or which ones are more
+worthwhile to optimize. What if we could have that feedback? Well we can; the
+:abbr:`FDO (Feedback-Directed Optimization)` technology makes this magic
+happen.
+
+Feedback-Directed Optimization
+==============================
+
+Traditional :abbr:`FDO (Feedback-Directed Optimization)` in GCC uses static
+instrumentation to collect edge and value profiles. GCC uses execution
+profiles consisting of basic block and edge frequency counts to guide the
+optimizations for things like instruction scheduling, basic block reordering,
+function splitting, and register allocation. According to Ramasamy, Yuan,
+Chen & Hundt, 2008, the current method of FDO in GCC involves the following
+steps:
+
+#. Build an instrumented version of the program for edge and value profiling,
+   the instrumentation build.
+
+#. Run the instrumented version with representative training data to collect
+   the execution profile. These runs typically incur significant overhead due
+   to the additional instrumentation code executed.
+
+#. Build an optimized version of the program by using the collected execution
+   profile to guide the optimizations, the FDO build.
+
+The instrumentation and FDO builds are tightly coupled. GCC requires both
+builds use the same in-line decisions and similar optimization flags to
+ensure that the :abbr:`CFG (Control-Flow Graph)` instrumented in the
+instrumentation build matches the CFG annotated with the profile data in the
+FDO build.
```

Comment:
No abbr is needed if the abbreviation was written out previously on the document or if abbr was used once before.